dbuf-lsp/src/common/pretty_printer.rs

```diff
@@ -0,0 +1,439 @@
+//! Provides utilities for pretty-printing AST (Abstract Syntax Tree) representations.
+//!
+//! This module contains tools to:
+//! - Format AST nodes as human-readable text
+//! - Track and update source locations during printing
+//!
+
+use std::io::{self, ErrorKind, Write};
+
+use std::rc::Rc;
+
+use dbuf_core::ast::operators::*;
+use dbuf_core::ast::parsed::definition::*;
+use dbuf_core::ast::parsed::*;
+use dbuf_core::location::*;
+
+type Str = String;
+type Loc = Location;
+
+// TODO:
+//   * make configuration
+//   * non mutable variant (?)
+//   * better use unsafe blocks instead of Rc recreation (?)
+//   * TODO resolutions in code
+
+/// A configurable AST pretty-printer that preserves source location information.
+///
+pub struct PrettyWriter<'a, W: Write> {
+    cursor: Position,
+    writer: &'a mut W,
+}
+
+impl<'a, W: Write> PrettyWriter<'a, W> {
+    const MESSAGE_TEXT: &'a str = "message ";
+    const ENUM_TEXT: &'a str = "enum ";
+
+    pub fn new(writer: &'a mut W) -> PrettyWriter<'a, W> {
+        PrettyWriter {
+            cursor: Position::new(0, 0),
+            writer,
+        }
+    }
+
+    fn new_line(&mut self) -> io::Result<()> {
+        self.cursor.line += 1;
+        self.cursor.character = 0;
+        writeln!(self.writer, "")?;
+        Ok(())
+    }
+
+    fn write(&mut self, s: &str) -> io::Result<()> {
+        self.cursor.character += s.len() as u32;
+        write!(self.writer, "{}", s)?;
+        Ok(())
+    }
+
+    fn write_tab(&mut self, len: usize) -> io::Result<()> {
+        self.cursor.character += len as u32;
+        let to_write = std::iter::repeat(" ").take(len).collect::<String>();
+        write!(self.writer, "{}", to_write)?;
+        Ok(())
+    }
+
+    pub fn parse_module(&mut self, module: &mut Module<Loc, Str>) -> io::Result<()> {
+        self.cursor = Position::new(0, 0);
+        let mut first = true;
+
+        for definition in module.iter_mut() {
+            if !first {
+                self.new_line()?;
+                self.new_line()?;
+            }
+            self.parse_type_definition(definition)?;
+            first = false;
+        }
+
+        Ok(())
+    }
+
+    // TODO: somewhere here parse location for name
+    fn parse_type_definition(
```

Comment:
Well, it's not called _parsing_, it's _pretty-printing_ as you had it in the module name